kotlin-inject-compiler/core/src/main/kotlin/me/tatarka/inject/compiler/TypeCollector.kt

```diff
@@ -164,6 +164,11 @@ class TypeCollector(private val provider: AstProvider, private val options: Opti
                             accessor = accessor + parameter.name,
                             typeInfo = elemTypeInfo
                         )
+
+                        // check for duplicates from previously collected Results
+                        for (parent in parents.minus(parentResult).plus(this)) {
+                            checkDuplicateTypesBetweenResults(provider, parentResult, parent)
+                        }
```

Comment:
Hm, I don't love that this is n^2. Wondering if it would be better to check once outside of the loop once all the parents have been collected?

Reply:
thanks for the feedback, I'll change that

Reply:
I haven't yet figured out a way to avoid the O(n^2) complexity. Perhaps memoization could be a solution. Realistically, how many parents can a component have before this is an issue?

One improvement I made is avoiding repetitive checks in recursive calls, which was an issue with deep hierarchies. By adding the check `!accessor.isNotEmpty()`, I aim to ensure checks are performed only once per component.

```kotlin
// When accessor is empty, check for duplicates from previously collected Results
if (!accessor.isNotEmpty()) {
    for (parent in parents.minus(parentResult).plus(this)) {
        checkDuplicateTypesBetweenResults(provider, parentResult, parent)
    }
}
```

Reply:
You are already collecting into a hashmap to look for duplicates, if you add the previous results each time you should only have to loop once.